exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -0,0 +1,241 @@
+        serviceId, services.get(serviceId), newService.getService());
+  }
+
+  /**
+   * Creates a TestKit network with a single validator node for a single service.
+   */
+  public static TestKit forService(Class<? extends ServiceModule> serviceModule) {
+    return new TestKit(singletonList(serviceModule), EmulatedNodeType.VALIDATOR, (short) 0, null);
+  }
+
+  /**
+   * Returns an instance of a service with the given service id and service class. Only
+   * user-defined services can be requested, i.e., it is not possible to get an instance of a
+   * built-in service such as the time oracle.
+   *
+   * @return the service instance or null if there is no service with such id
+   * @throws IllegalArgumentException if the service with given id was not found or could not be
+   *     cast to given class
+   */
+  public <T extends Service> T getService(short serviceId, Class<T> serviceClass) {
+    Service service = services.get(serviceId);
+    checkArgument(service != null, "Service with given id=%s was not found", serviceId);
+    checkArgument(service.getClass().equals(serviceClass),
+        "Service (id=%s, class=%s) cannot be cast to %s",
+        serviceId, service.getClass().getCanonicalName(), serviceClass.getCanonicalName());
+    return serviceClass.cast(service);
+  }
+
+  private native long nativeCreateTestKit(UserServiceAdapter[] services, boolean auditor,
+      short withValidatorCount, TimeProvider timeProvider);
+
+  private native long nativeCreateSnapshot(long nativeHandle);
+
+  private native Block nativeCreateBlock(long nativeHandle);
+
+  private native Block nativeCreateBlockWithTransactions(long nativeHandle, byte[][] transactions);
+
+  private native EmulatedNode nativeGetEmulatedNode(long nativeHandle);
+
+  /**
+   * Creates a new builder for the TestKit.
+   *
+   * @param nodeType type of the main TestKit node - either validator or auditor. Note that
+   * {@link Service#afterCommit(BlockCommittedEvent)} logic will only be called on the main TestKit
+   *     node of this type
+   */
+  public static Builder builder(EmulatedNodeType nodeType) {
+    checkNotNull(nodeType);
+    return new Builder(nodeType);
+  }
+
+  /**
+   * Builder for the TestKit.
+   */
+  public static final class Builder {
+
+    private EmulatedNodeType nodeType;
+    private short validatorCount;
+    private List<Class<? extends ServiceModule>> services = new ArrayList<>();
+    private TimeProvider timeProvider;
+
+    private Builder(EmulatedNodeType nodeType) {
+      // TestKit network should have at least one validator node
+      if (nodeType == EmulatedNodeType.AUDITOR) {
+        validatorCount = 1;
+      }
+      this.nodeType = nodeType;
+    }
+
+    /**
+     * Sets number of additional validator nodes in the TestKit network. Note that
+     * regardless of the configured number of validators, only a single service will be
+     * instantiated.
+     */
+    public final Builder withValidators(short validatorCount) {
```

Comment:
`final` is redundant for methods such as the class is final.

Reply:
Yep, I think it is only required (even in final classes) for @SafeVarargs methods.

Reply:
Yes, it was needed for @SafeVarargs method, so I added others for consistency.